--- "a/htmlPreviews\\htmlPreview9.html"
+++ "b/htmlPreviews\\htmlPreview10.html"
@@ -311,27 +311,43 @@
 		let cube2 = new Cube(new Vector3(1, 0, 2.5), new Vector3(0, 0, 15), new Vector3(1, 0.7, 0.7));
 		cube2.pos = Vector3.sub(cube2.pos, new Vector3(1.6, 0, 0));
 
-		let allTris = [];
-		allTris.push(...cube1.getTransformedTriangles());
-		allTris.push(...cube2.getTransformedTriangles());
-
-		for (let tri of allTris) {
-			let vertices = tri.vertices;
-			for (let i = 0; i < vertices.length; i++) {
-				vertices[i].mulMat4x4(projectionMatrix);
-				vertices[i] = Vector3.div(vertices[i], vertices[i].w);
-				vertices[i] = new Vector3(
-					vertices[i].x * canvas.width + canvas.width / 2,
-					-vertices[i].y * canvas.height + canvas.height / 2,
-					1);
+		let lastTime = performance.now();
+		let dt = 0;
+
+		function update(time = performance.now()) {
+			dt = (time - lastTime) / 1000;
+			lastTime = time;
+
+			ctx.clearRect(0, 0, canvas.width, canvas.height);
+			cube1.rotateY(50 * dt);
+			cube2.rotateX(50 * dt);
+
+			let allTris = [];
+			allTris.push(...cube1.getTransformedTriangles());
+			allTris.push(...cube2.getTransformedTriangles());
+			for (let tri of allTris) {
+				let vertices = tri.vertices;
+				for (let i = 0; i < vertices.length; i++) {
+					vertices[i].mulMat4x4(projectionMatrix);
+					vertices[i] = Vector3.div(vertices[i], vertices[i].w);
+					vertices[i] = new Vector3(
+						vertices[i].x * canvas.width + canvas.width / 2,
+						-vertices[i].y * canvas.height + canvas.height / 2,
+						1);
+				}
+				ctx.beginPath();
+				ctx.moveTo(vertices[0].x, vertices[0].y);
+				ctx.lineTo(vertices[1].x, vertices[1].y);
+				ctx.lineTo(vertices[2].x, vertices[2].y);
+				ctx.closePath();
+				ctx.stroke();
 			}
-			ctx.beginPath();
-			ctx.moveTo(vertices[0].x, vertices[0].y);
-			ctx.lineTo(vertices[1].x, vertices[1].y);
-			ctx.lineTo(vertices[2].x, vertices[2].y);
-			ctx.closePath();
-			ctx.stroke();
+
+			requestAnimationFrame(update);
 		}
+
+		update();
+
 	</script>
 </body>
 
