--- "a/htmlPreviews\\htmlPreview27.html"
+++ "b/htmlPreviews\\htmlPreview28.html"
@@ -676,6 +676,12 @@
 						tri.vertices[0],
 						new Vector3(bbox[0], bbox[2], 0));
 
+					let perspectiveWorldVertices = [
+						Vector3.mul(tri.worldVertices[0], tri.uv[0].w),
+						Vector3.mul(tri.worldVertices[1], tri.uv[1].w),
+						Vector3.mul(tri.worldVertices[2], tri.uv[2].w)
+					];
+
 					for (let y = bbox[2]; y <= bbox[3]; y++) {
 						let edgeFunction01 = edgeFunctionRow01;
 						let edgeFunction12 = edgeFunctionRow12;
@@ -690,11 +696,14 @@
 								if (z < this.depthBuffer[y][x]) {
 									this.depthBuffer[y][x] = z;
 
+									let pixelInvZ = tri.uv[0].w * baryCoord0 + tri.uv[1].w * baryCoord1 + tri.uv[2].w * baryCoord2;
+
 									let pixelWorldPos = new Vector3(
-										tri.worldVertices[0].x * baryCoord0 + tri.worldVertices[1].x * baryCoord1 + tri.worldVertices[2].x * baryCoord2,
-										tri.worldVertices[0].y * baryCoord0 + tri.worldVertices[1].y * baryCoord1 + tri.worldVertices[2].y * baryCoord2,
-										tri.worldVertices[0].z * baryCoord0 + tri.worldVertices[1].z * baryCoord1 + tri.worldVertices[2].z * baryCoord2
+										(perspectiveWorldVertices[0].x * baryCoord0 + perspectiveWorldVertices[1].x * baryCoord1 + perspectiveWorldVertices[2].x * baryCoord2),
+										(perspectiveWorldVertices[0].y * baryCoord0 + perspectiveWorldVertices[1].y * baryCoord1 + perspectiveWorldVertices[2].y * baryCoord2),
+										(perspectiveWorldVertices[0].z * baryCoord0 + perspectiveWorldVertices[1].z * baryCoord1 + perspectiveWorldVertices[2].z * baryCoord2)
 									);
+									pixelWorldPos = Vector3.div(pixelWorldPos, pixelInvZ);
 
 									let pointLightVec = Vector3.sub(pixelWorldPos, lightCube.pos);
 									let distance = pointLightVec.magnitude();
@@ -710,7 +719,7 @@
 										let pixelUV = new Vector2(
 											tri.uv[0].u * baryCoord0 + tri.uv[1].u * baryCoord1 + tri.uv[2].u * baryCoord2,
 											tri.uv[0].v * baryCoord0 + tri.uv[1].v * baryCoord1 + tri.uv[2].v * baryCoord2,
-											tri.uv[0].w * baryCoord0 + tri.uv[1].w * baryCoord1 + tri.uv[2].w * baryCoord2
+											pixelInvZ
 										);
 										pixelUV = Vector2.div(pixelUV, pixelUV.w);
 
@@ -855,16 +864,7 @@
 				tri.perspectiveDivide();
 			}
 
-			let clippedTris = visibleTris;
-			for (let plane of edgePlanes) {
-				let newTris = [];
-				for (let tri of clippedTris) {
-					newTris.push(...tri.clipAgainstPlane(plane));
-				}
-				clippedTris = newTris;
-			}
-
-			rasterizer.processTriangles(clippedTris);
+			rasterizer.processTriangles(visibleTris);
 			rasterizer.drawCall();
 
 			requestAnimationFrame(update);
