--- "a/htmlPreviews\\htmlPreview22.html"
+++ "b/htmlPreviews\\htmlPreview23.html"
@@ -329,18 +329,6 @@
 				return outTris;
 			}
 
-			isPointInTriangle(p) {
-				const v0 = [this.vertices[0].x, this.vertices[0].y];
-				const v1 = [this.vertices[1].x, this.vertices[1].y];
-				const v2 = [this.vertices[2].x, this.vertices[2].y];
-
-				const e1 = edgeFunction(v0, v1, p);
-				const e2 = edgeFunction(v1, v2, p);
-				const e3 = edgeFunction(v2, v0, p);
-
-				return e1 > 0 && e2 > 0 && e3 > 0;
-			}
-
 			boundingBox() {
 				let minX = Math.min(this.vertices[0].x, this.vertices[1].x, this.vertices[2].x);
 				let maxX = Math.max(this.vertices[0].x, this.vertices[1].x, this.vertices[2].x);
@@ -355,16 +343,8 @@
 				return [minX, maxX, minY, maxY];
 			}
 
-			barycentricCoords(p) {
-				const v0 = [this.vertices[0].x, this.vertices[0].y];
-				const v1 = [this.vertices[1].x, this.vertices[1].y];
-				const v2 = [this.vertices[2].x, this.vertices[2].y];
-
-				const area = edgeFunction(v0, v1, v2);
-				const w0 = edgeFunction(v1, v2, p) / area;
-				const w1 = edgeFunction(v2, v0, p) / area;
-				const w2 = edgeFunction(v0, v1, p) / area;
-				return [w0, w1, w2];
+			calculateSignedDoubleArea() {
+				this.signedDoubleArea = edgeFunction(this.vertices[0], this.vertices[1], this.vertices[2]);
 			}
 		}
 
@@ -584,19 +564,46 @@
 							vertices[i].z);
 					}
 
+					tri.calculateSignedDoubleArea();
+
 					const bbox = tri.boundingBox();
+
+					let edgeFunctionRow01 = edgeFunction(
+						tri.vertices[0],
+						tri.vertices[1],
+						new Vector3(bbox[0], bbox[2], 0));
+					let edgeFunctionRow12 = edgeFunction(
+						tri.vertices[1],
+						tri.vertices[2],
+						new Vector3(bbox[0], bbox[2], 0));
+					let edgeFunctionRow20 = edgeFunction(
+						tri.vertices[2],
+						tri.vertices[0],
+						new Vector3(bbox[0], bbox[2], 0));
+
 					for (let y = bbox[2]; y <= bbox[3]; y++) {
+						let edgeFunction01 = edgeFunctionRow01;
+						let edgeFunction12 = edgeFunctionRow12;
+						let edgeFunction20 = edgeFunctionRow20;
 						for (let x = bbox[0]; x <= bbox[1]; x++) {
-							if (tri.isPointInTriangle([x, y])) {
-								const baryCoords = tri.barycentricCoords([x, y]);
-								const z = tri.vertices[0].z * baryCoords[0] + tri.vertices[1].z * baryCoords[1] + tri.vertices[2].z * baryCoords[2];
+							if (edgeFunction01 > 0 && edgeFunction12 > 0 && edgeFunction20 > 0) {
+								const baryCoord0 = edgeFunction12 / tri.signedDoubleArea;
+								const baryCoord1 = edgeFunction20 / tri.signedDoubleArea;
+								const baryCoord2 = edgeFunction01 / tri.signedDoubleArea;
+								const z = tri.vertices[0].z * baryCoord0 + tri.vertices[1].z * baryCoord1 + tri.vertices[2].z * baryCoord2;
 
 								if (z < this.depthBuffer[y][x]) {
 									this.depthBuffer[y][x] = z;
 									this.screenBuffer[y][x].updateColor(tri.color.r, tri.color.g, tri.color.b);
 								}
 							}
+							edgeFunction01 += (tri.vertices[0].y - tri.vertices[1].y);
+							edgeFunction12 += (tri.vertices[1].y - tri.vertices[2].y);
+							edgeFunction20 += (tri.vertices[2].y - tri.vertices[0].y);
 						}
+						edgeFunctionRow01 += (tri.vertices[1].x - tri.vertices[0].x);
+						edgeFunctionRow12 += (tri.vertices[2].x - tri.vertices[1].x);
+						edgeFunctionRow20 += (tri.vertices[0].x - tri.vertices[2].x);
 					}
 				}
 			}
@@ -616,9 +623,9 @@
 				ctx.putImageData(this.imageData, 0, 0);
 			}
 		}
-		function edgeFunction(a, b, p) {
-            return (a[1] - b[1]) * (p[0] - a[0]) + (b[0] - a[0]) * (p[1] - a[1]);
-        }
+		function edgeFunction(v0, v1, p) {
+			return (v0.y - v1.y) * (p.x - v0.x) + (v1.x - v0.x) * (p.y - v0.y);
+		}
 
 		const canvas = document.getElementById("gameCanvas");
 		const ctx = canvas.getContext("2d");
