--- "a/htmlPreviews\\htmlPreview16.html"
+++ "b/htmlPreviews\\htmlPreview17.html"
@@ -496,6 +496,12 @@
 		let projectionMatrix = Mat4x4.Projection(fov, canvas.height, canvas.width, znear, zfar);
 		let camera = new Camera(Vector3.zero, Vector3.zero, 2, 0.2);
 		let nearPlane = new Plane(Vector3.forward, new Vector3(0, 0, znear));
+		let edgePlanes = [
+			new Plane(Vector3.down, new Vector3(0, 0.5, 0)),
+			new Plane(Vector3.up, new Vector3(0, -0.5, 0)),
+			new Plane(Vector3.right, new Vector3(-0.5, 0, 0)),
+			new Plane(Vector3.left, new Vector3(0.5, 0, 0))
+		];
 		let keyStates = [];
 
 		const fpsInfo = document.getElementById("fps");
@@ -548,7 +554,18 @@
 			for (let tri of visibleTris) {
 				tri.mulMat4x4(projectionMatrix);
 				tri.perspectiveDivide();
+			}
+
+			let clippedTris = visibleTris;
+			for (let plane of edgePlanes) {
+				let newTris = [];
+				for (let tri of clippedTris) {
+					newTris.push(...tri.clipAgainstPlane(plane));
+				}
+				clippedTris = newTris;
+			}
 
+			for (let tri of clippedTris) {
 				let vertices = tri.vertices;
 				for (let i = 0; i < vertices.length; i++) {
 					vertices[i] = new Vector3(
