--- "a/htmlPreviews\\htmlPreview21.html"
+++ "b/htmlPreviews\\htmlPreview22.html"
@@ -337,6 +337,7 @@
 				const e1 = edgeFunction(v0, v1, p);
 				const e2 = edgeFunction(v1, v2, p);
 				const e3 = edgeFunction(v2, v0, p);
+
 				return e1 < 0 && e2 < 0 && e3 < 0;
 			}
 
@@ -353,6 +354,18 @@
 
 				return [minX, maxX, minY, maxY];
 			}
+
+			barycentricCoords(p) {
+				const v0 = [this.vertices[0].x, this.vertices[0].y];
+				const v1 = [this.vertices[1].x, this.vertices[1].y];
+				const v2 = [this.vertices[2].x, this.vertices[2].y];
+
+				const area = edgeFunction(v0, v1, v2);
+				const w0 = edgeFunction(v1, v2, p) / area;
+				const w1 = edgeFunction(v2, v0, p) / area;
+				const w2 = edgeFunction(v0, v1, p) / area;
+				return [w0, w1, w2];
+			}
 		}
 
 		class Plane {
@@ -532,13 +545,17 @@
 			constructor() {
 				this.imageData = ctx.createImageData(canvas.width, canvas.height);
 				this.screenBuffer = [];
+				this.depthBuffer = [];
 
 				for (let i = 0; i < canvas.height; i++) {
 					let screenRow = [];
+					let depthRow = [];
 					for (let j = 0; j < canvas.width; j++) {
 						screenRow.push(new Color(255, 255, 255));
+						depthRow.push(Infinity);
 					}
 					this.screenBuffer.push(screenRow);
+					this.depthBuffer.push(depthRow);
 				}
 			}
 
@@ -546,6 +563,7 @@
 				for (let i = 0; i < this.screenBuffer.length; i++) {
 					for (let j = 0; j < this.screenBuffer[i].length; j++) {
 						this.screenBuffer[i][j] = new Color(240, 240, 240);
+						this.depthBuffer[i][j] = Infinity;
 					}
 				}
 			}
@@ -561,14 +579,20 @@
 						vertices[i] = new Vector3(
 							vertices[i].x * canvas.width + canvas.width / 2,
 							-vertices[i].y * canvas.height + canvas.height / 2,
-							1);
+							vertices[i].z);
 					}
 
 					const bbox = tri.boundingBox();
 					for (let y = bbox[2]; y <= bbox[3]; y++) {
 						for (let x = bbox[0]; x <= bbox[1]; x++) {
 							if (tri.isPointInTriangle([x, y])) {
-								this.placePixel([x, y], tri.color);
+								const baryCoords = tri.barycentricCoords([x, y]);
+								const z = tri.vertices[0].z * baryCoords[0] + tri.vertices[1].z * baryCoords[1] + tri.vertices[2].z * baryCoords[2];
+
+								if (z < this.depthBuffer[y][x]) {
+									this.depthBuffer[y][x] = z;
+									this.placePixel([x, y], tri.color);
+								}
 							}
 						}
 					}
@@ -649,11 +673,6 @@
 					visibleTris.push(...tri.clipAgainstPlane(nearPlane));
 				}
 			}
-			visibleTris.sort((a, b) => {
-				let distA = a.getCenter().magnitude();
-				let distB = b.getCenter().magnitude();
-				return distB - distA;
-			});
 
 			for (let tri of visibleTris) {
 				tri.mulMat4x4(projectionMatrix);
