--- "a/htmlPreviews\\htmlPreview6.html"
+++ "b/htmlPreviews\\htmlPreview7.html"
@@ -142,15 +142,17 @@
 
 		const canvas = document.getElementById("gameCanvas");
 		const ctx = canvas.getContext("2d");
+		const fov = 60;
+		let fovMultiplier = 1.0 / Math.tan(fov * 0.5 / 180.0 * Math.PI);
 
-		let cube1 = makeCube(new Vector3(0.7, -0.7, 2), new Vector3(1, 1, 1));
+		let cube1 = makeCube(new Vector3(0.7, -0.7, 3), new Vector3(1, 1, 1));
 		for (let tri of cube1) {
 			vertices = tri.vertices;
 			for (let i = 0; i < vertices.length; i++) {
 				let v = Vector3.div(vertices[i], vertices[i].z);
 				vertices[i] = new Vector3(
-					v.x * (canvas.height / 2) + canvas.width / 2,
-					-v.y * (canvas.height / 2) + canvas.height / 2,
+					v.x * fovMultiplier * (canvas.height / 2) + canvas.width / 2,
+					-v.y * fovMultiplier * (canvas.height / 2) + canvas.height / 2,
 					1);
 			}
 			ctx.beginPath();
