--- "a/htmlPreviews\\htmlPreview33.html"
+++ "b/htmlPreviews\\htmlPreview34.html"
@@ -15,11 +15,20 @@
 		<input type="range" id="dirIntensity" min="0" max="5" step="0.1" value="1" />
 		<span id="dirIntensityText" class="slider-value">1.0</span>
 	</div>
+	<div>
+		Directional light rotate speed:
+		<input type="range" id="dirRotateSpeed" min="0" max="50" step="0.1" value="10" />
+		<span id="dirRotateSpeedText" class="slider-value">10</span>
+	</div>
 	<div>
 		Point light intensity:
 		<input type="range" id="pointIntensity" min="0" max="5" step="0.1" value="1" />
 		<span id="pointIntensityText" class="slider-value">1.0</span>
 	</div>
+	<div>
+		Shadow:
+		<input type="checkbox" id="useShadow" checked="checked" />
+	</div>
 	<script>
 		class Vector2 {
 			constructor(u = 0, v = 0, w = 1) {
@@ -754,7 +763,8 @@
 				let up = Math.abs(Vector3.dot(this.dir, Vector3.up)) < 0.99 ? Vector3.up : Vector3.forward;
 				let right = Vector3.cross(up, this.dir).normalize();
 				up = Vector3.cross(this.dir, right).normalize();
-				let frustumCenter = camera.getFrustumCenter(shadowMapDistance);
+				//let frustumCenter = camera.getFrustumCenter(shadowMapDistance);
+				let frustumCenter = Vector3.zero;
 				let lightPos = Vector3.sub(frustumCenter, Vector3.mul(this.dir, shadowMapDistance));
 				this.viewMatrix = Mat4x4.View(right, up, this.dir, lightPos);
 				this.projectionMatrix = Mat4x4.Orthographic(shadowMapDistance, 0.1, shadowMapDistance);
@@ -923,7 +933,7 @@
 									);
 									pixelWorldPos = Vector3.div(pixelWorldPos, pixelInvZ);
 
-									let inShadow = dirLight.isInShadow(pixelWorldPos, triWorldNormal);
+									let inShadow = useShadow ? dirLight.isInShadow(pixelWorldPos, triWorldNormal) : false;
 									let shadowFactor = inShadow ? 0.5 : 1.0;
 
 									let pointLightVec = Vector3.sub(pixelWorldPos, lightCube.pos);
@@ -987,7 +997,7 @@
 		const fov = 60;
 		let znear = 0.01, zfar = 1000;
 		let projectionMatrix = Mat4x4.Projection(fov, canvas.height, canvas.width, znear, zfar);
-		let camera = new Camera(new Vector3(0, 1.5, 1.5), new Vector3(-15, 0, 0), 2, 0.2);
+		let camera = new Camera(new Vector3(0, 10, -10), new Vector3(-25, 0, 0), 2, 0.2);
 		let nearPlane = new Plane(Vector3.forward, new Vector3(0, 0, znear));
 		let edgePlanes = [
 			new Plane(Vector3.down, new Vector3(0, 1, 0)),
@@ -1002,26 +1012,31 @@
 
 		const lightDirIntensitySlider = document.getElementById("dirIntensity");
 		const lightDirIntensityText = document.getElementById("dirIntensityText");
+		const lightDirRotateSpeedSlider = document.getElementById("dirRotateSpeed");
+		const lightDirRotateSpeedText = document.getElementById("dirRotateSpeedText");
+		const shadowCheckbox = document.getElementById("useShadow");
+		let useShadow = true;
 		let shadowMapSize = 512;
 		let shadowBias = 0.01;
-		let shadowMapDistance = 5;
-		let dirLight = new DirectionalLight(new Vector3(-1, -1, -1), parseFloat(lightDirIntensitySlider.value));
+		let shadowMapDistance = 20;
+		let dirLightDefaultDir = new Vector3(-1, -1, -1);
+		let dirLight = new DirectionalLight(dirLightDefaultDir, parseFloat(lightDirIntensitySlider.value));
+		let dirLightRotation = Quaternion.buildQuaternionEuler(Vector3.zero);
+		let dirLightRotateSpeed = parseFloat(lightDirRotateSpeedSlider.value);
 
 		const pointLightIntensitySlider = document.getElementById("pointIntensity");
 		const pointLightIntensityText = document.getElementById("pointIntensityText");
 		let pointLightIntensity = parseFloat(pointLightIntensitySlider.value);
 
 		let brickTexture = new Texture("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/images/textures/brick.jpg");
-		let grassTexture = new Texture("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/images/textures/grass.jpg");
+		let woodTexture = new Texture("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/images/textures/Wood_Tower_Col.jpg");
 
 		let cubeMesh = new Cube();
-		let cube1 = new GameObject(new Vector3(0, 0.35, 5), new Vector3(0, 0, 0), new Vector3(0.7, 0.7, 0.7), new Color(255, 0, 0), cubeMesh, brickTexture);
-		let cube2 = new GameObject(new Vector3(0, 0, 5), Vector3.zero, new Vector3(3, 0.1, 3), new Color(0, 255, 0), cubeMesh, brickTexture);
-		let lightCube = new GameObject(new Vector3(0, 2, 5), Vector3.zero, new Vector3(0.1, 0.1, 0.1), new Color(255, 255, 255), cubeMesh, brickTexture);
-
-		let loadedObj = new GameObject(new Vector3(0, 0, 10), Vector3.zero, Vector3.one, new Color(255, 255, 255),
-			new ObjMesh("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/models/grass_block.obj"),
-			grassTexture);
+		let woodTower = new GameObject(new Vector3(0, 0, 0), Vector3.zero, new Vector3(1, 1, 1), new Color(255, 255, 255),
+			new ObjMesh("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/models/wooden watch tower2.obj"),
+			woodTexture);
+		let ground = new GameObject(new Vector3(0, 0, 0), Vector3.zero, new Vector3(30, 1, 30), new Color(0, 255, 0), cubeMesh, brickTexture);
+		let lightCube = new GameObject(new Vector3(0, 2, 5), Vector3.zero, new Vector3(0.1, 0.1, 0.1), new Color(255, 255, 255), cubeMesh);
 
 		let lastTime = performance.now();
 		let dt = 0;
@@ -1043,7 +1058,10 @@
 			camera.updateMovement(dt, keyStates);
 			const viewMat = camera.getViewMatrix();
 
-			cube1.rotate(0, 50 * dt, 0);
+			let q = Quaternion.buildQuaternionEuler(new Vector3(0, dt * dirLightRotateSpeed, 0));
+			dirLightRotation = Quaternion.multiply(q, dirLightRotation);
+			dirLight.dir = dirLightRotation.rotateVector(dirLightDefaultDir);
+
 			if (keyStates['i']) {
 				lightCube.pos.z += dt * 2;
 			} else if (keyStates['k']) {
@@ -1059,13 +1077,14 @@
 			}
 
 			let objectTris = [];
-			objectTris.push(...cube1.getTransformedTriangles());
-			objectTris.push(...cube2.getTransformedTriangles());
+			objectTris.push(...ground.getTransformedTriangles());
 			objectTris.push(...lightCube.getTransformedTriangles());
-			objectTris.push(...loadedObj.getTransformedTriangles());
+			objectTris.push(...woodTower.getTransformedTriangles());
 
-			dirLight.buildViewAndProjectionMatrix();
-			dirLight.buildShadowMap(objectTris);
+			if (useShadow) {
+				dirLight.buildViewAndProjectionMatrix();
+				dirLight.buildShadowMap(objectTris);
+			}
 
 			let visibleTris = [];
 			for (let tri of objectTris) {
@@ -1093,11 +1112,20 @@
 			lightDirIntensityText.textContent = dirLight.intensity.toFixed(2);
 		});
 
+		lightDirRotateSpeedSlider.addEventListener("input", () => {
+			dirLightRotateSpeed = parseFloat(lightDirRotateSpeedSlider.value);
+			lightDirRotateSpeedText.textContent = dirLightRotateSpeed.toFixed(2);
+		});
+
 		pointLightIntensitySlider.addEventListener("input", () => {
 			pointLightIntensity = parseFloat(pointLightIntensitySlider.value);
 			pointLightIntensityText.textContent = pointLightIntensity.toFixed(2);
 		});
 
+		shadowCheckbox.addEventListener("change", () => {
+			useShadow = shadowCheckbox.checked;
+		});
+
 		canvas.addEventListener("click", async () => {
 			await canvas.requestPointerLock({
 				unadjustedMovement: true,
