--- "a/htmlPreviews\\htmlPreview10.html"
+++ "b/htmlPreviews\\htmlPreview11.html"
@@ -9,6 +9,7 @@
 <body style="text-align: center; background: #f0f0f0">
 	<h1>3D Engine</h1>
 	<canvas id="gameCanvas" width="400" height="300" style="border: 2px solid #aaa; background: #fff"></canvas>
+	<h2 id="fps">FPS: 0</h2>
 	<script>
 		class Vector3 {
 			constructor(x = 0, y = 0, z = 0, w = 1) {
@@ -307,6 +308,9 @@
 		let znear = 0.01, zfar = 1000;
 		let projectionMatrix = Mat4x4.Projection(fov, canvas.height, canvas.width, znear, zfar);
 
+		const fpsInfo = document.getElementById("fps");
+		let frames = 0, fpsTimeCounter = 0;
+
 		let cube1 = new Cube(new Vector3(0.6, -0.7, 2.5), new Vector3(45, 0, 0), new Vector3(0.7, 0.7, 0.7));
 		let cube2 = new Cube(new Vector3(1, 0, 2.5), new Vector3(0, 0, 15), new Vector3(1, 0.7, 0.7));
 		cube2.pos = Vector3.sub(cube2.pos, new Vector3(1.6, 0, 0));
@@ -318,6 +322,14 @@
 			dt = (time - lastTime) / 1000;
 			lastTime = time;
 
+			fpsTimeCounter += dt;
+			frames++;
+			if (fpsTimeCounter >= 1) {
+				fpsInfo.textContent = `FPS: ${frames}`;
+				frames = 0;
+				fpsTimeCounter = 0;
+			}
+
 			ctx.clearRect(0, 0, canvas.width, canvas.height);
 			cube1.rotateY(50 * dt);
 			cube2.rotateX(50 * dt);
