--- "a/htmlPreviews\\htmlPreview3.html"
+++ "b/htmlPreviews\\htmlPreview4.html"
@@ -15,6 +15,7 @@
         <input type="range" id="dirIntensity" min="0" max="5" step="0.1" value="1" />
         <span id="dirIntensityText" class="slider-value">1.0</span>
     </div>
+    <canvas id="portal1Canvas" width="400" height="300" style="border: 2px solid #aaa; background: #fff"></canvas>
     <script>
         class Vector2 {
             constructor(u = 0, v = 0, w = 1) {
@@ -250,6 +251,12 @@
                     res.c,
                     res.d);
             }
+
+            inverseRotateVector(v) {
+                const qVec = new Quaternion(0, v.x, v.y, v.z);
+                const res = Quaternion.multiply(Quaternion.multiply(this.conjugate(), qVec), this);
+                return new Vector3(res.b, res.c, res.d);
+            }
         }
 
         class Triangle {
@@ -910,7 +917,10 @@
                 }
             }
 
-            drawCall() {
+            drawCall(canvasCtx = null) {
+                if (canvasCtx == null)
+                    canvasCtx = ctx;
+
                 let index = 0;
                 for (let y = 0; y < canvas.height; y++) {
                     for (let x = 0; x < canvas.width; x++) {
@@ -922,15 +932,84 @@
                     }
                 }
 
-                ctx.putImageData(this.imageData, 0, 0);
+                canvasCtx.putImageData(this.imageData, 0, 0);
             }
         }
         function edgeFunction(v0, v1, p) {
             return (v0.y - v1.y) * (p.x - v0.x) + (v1.x - v0.x) * (p.y - v0.y);
         }
 
+        class Portal {
+            constructor(name, pos, eulerAngles, scale, color) {
+                this.name = name;
+                this.pos = pos;
+                this.rotation = Quaternion.buildQuaternionEuler(eulerAngles);
+                this.scale = scale;
+
+                let portalQuadMesh = new Quad();
+                this.quad = new GameObject(pos, eulerAngles, scale, color, portalQuadMesh, null);
+                this.rasterizer = new Rasterizer();
+                this.camera = new Camera(pos, eulerAngles, 8, 0.2);
+            }
+            setOtherPortal(otherPortal) {
+                this.otherPortal = otherPortal;
+            }
+
+            runRenderPipeline(objectTris) {
+                let tris = [];
+                for (let tri of objectTris) {
+                    let cloneTri = tri.clone();
+                    tris.push(cloneTri);
+                }
+
+                this.rasterizer.clearScreen();
+
+                let viewMatPortal = this.camera.getViewMatrix();
+
+                let visibleTris = [];
+                for (let tri of tris) {
+                    if (Vector3.dot(tri.getNormal(), Vector3.sub(tri.getCenter(), this.camera.pos)) < 0) {
+                        tri.mulMat4x4(viewMatPortal);
+                        visibleTris.push(...tri.clipAgainstPlane(nearPlane));
+                    }
+                }
+
+                for (let tri of visibleTris) {
+                    tri.mulMat4x4(projectionMatrix);
+                    tri.perspectiveDivide();
+                }
+
+                this.rasterizer.processTriangles(visibleTris);
+            }
+
+            calculateTransformToOtherPortalTransform(position, rotation) {
+                let otherPortalUp = this.otherPortal.rotation.rotateVector(Vector3.up).normalize();
+                let rotate180 = Quaternion.buildQuaternionAxisAngle(otherPortalUp, 180);
+
+                let worldPositionRelativeToThisPortal = Vector3.sub(position, this.pos);
+                let localPositionRelativeToThisPortal = this.rotation.inverseRotateVector(worldPositionRelativeToThisPortal);
+                let worldPositionRelativeToOtherPortal = this.otherPortal.rotation.rotateVector(localPositionRelativeToThisPortal);
+                let newPos = Vector3.add(this.otherPortal.pos, rotate180.rotateVector(worldPositionRelativeToOtherPortal));
+
+                let thisPortalInverseRotation = this.rotation.conjugate();
+                let localRotationRelativeToThisPortal = Quaternion.multiply(thisPortalInverseRotation, rotation);
+                let worldRotationRelativeToOtherPortal = Quaternion.multiply(this.otherPortal.rotation, localRotationRelativeToThisPortal);
+                let newRotation = Quaternion.multiply(rotate180, worldRotationRelativeToOtherPortal);
+
+                return [newPos, newRotation];
+            }
+
+            updatePortalCameraBaseOnPlayer(playerCameraPos, playerCameraRotation) {
+                let [cameraPos, cameraRotation] = this.calculateTransformToOtherPortalTransform(playerCameraPos, playerCameraRotation);
+                this.camera.pos = cameraPos;
+                this.camera.rotation = cameraRotation;
+            }
+        }
+
         const canvas = document.getElementById("gameCanvas");
+        const portal1Canvas = document.getElementById("portal1Canvas");
         const ctx = canvas.getContext("2d");
+        const portal1Ctx = portal1Canvas.getContext("2d");
         const rasterizer = new Rasterizer();
         const fov = 60;
         let znear = 0.01, zfar = 1000;
@@ -953,13 +1032,17 @@
         let dirLightDefaultDir = new Vector3(0.2, -1, 0);
         let dirLight = new DirectionalLight(dirLightDefaultDir, parseFloat(lightDirIntensitySlider.value));
 
+        let playerCubeTexture = new Texture("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/images/textures/wheatly.png");
         let brickTexture = new Texture("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/images/textures/brick.jpg");
 
-        let quadMesh = new Quad();
         let cubeMesh = new Cube();
         let ground = new GameObject(new Vector3(0, 0, 0), Vector3.zero, new Vector3(30, 1, 30), new Color(0, 255, 0), cubeMesh, brickTexture);
-        let portal1 = new GameObject(new Vector3(-3, 3, 0), new Vector3(0, 180, 0), new Vector3(3, 6, 3), new Color(255, 201, 94), quadMesh);
-        let portal2 = new GameObject(new Vector3(3, 3, 0), new Vector3(0, 180, 0), new Vector3(3, 6, 3), new Color(17, 155, 255), quadMesh);
+        let playerCube = new GameObject(camera.pos, camera.rotation, new Vector3(0.5, 0.5, 0.5), new Color(255, 255, 255),
+            new ObjMesh("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/models/grass_block.obj"), playerCubeTexture);
+        let portal1 = new Portal("Portal 1", new Vector3(-3, 3, 0), new Vector3(0, 180, 0), new Vector3(3, 6, 3), new Color(255, 201, 94));
+        let portal2 = new Portal("Portal 2", new Vector3(3, 3, 0), new Vector3(0, 180, 0), new Vector3(3, 6, 3), new Color(17, 155, 255));
+        portal1.setOtherPortal(portal2);
+        portal2.setOtherPortal(portal1);
 
         let lastTime = performance.now();
         let dt = 0;
@@ -976,15 +1059,26 @@
                 fpsTimeCounter = 0;
             }
 
+            playerCube.pos = camera.pos;
+            playerCube.rotation = camera.rotation;
+
             rasterizer.clearScreen();
 
             camera.updateMovement(dt, keyStates);
+
+            portal1.updatePortalCameraBaseOnPlayer(camera.pos, camera.rotation);
+            portal2.updatePortalCameraBaseOnPlayer(camera.pos, camera.rotation);
+
             const viewMat = camera.getViewMatrix();
 
             let objectTris = [];
             objectTris.push(...ground.getTransformedTriangles());
-            objectTris.push(...portal1.getTransformedTriangles());
-            objectTris.push(...portal2.getTransformedTriangles());
+            objectTris.push(...playerCube.getTransformedTriangles());
+            objectTris.push(...portal1.quad.getTransformedTriangles());
+            objectTris.push(...portal2.quad.getTransformedTriangles());
+
+            portal1.runRenderPipeline(objectTris);
+            portal2.runRenderPipeline(objectTris);
 
             let visibleTris = [];
             for (let tri of objectTris) {
@@ -999,6 +1093,8 @@
                 tri.perspectiveDivide();
             }
 
+            portal1.rasterizer.drawCall(portal1Ctx);
+
             rasterizer.processTriangles(visibleTris);
             rasterizer.drawCall();
 
