--- "a/htmlPreviews\\htmlPreview9.html"
+++ "b/htmlPreviews\\htmlPreview10.html"
@@ -15,6 +15,10 @@
         <input type="range" id="dirIntensity" min="0" max="5" step="0.1" value="1" />
         <span id="dirIntensityText" class="slider-value">1.0</span>
     </div>
+    <div>
+        Controlling object:
+        <span id="controllingObject">Orange portal</span>
+    </div>
     <canvas id="portal1Canvas" width="400" height="300" style="border: 2px solid #aaa; background: #fff"></canvas>
     <script>
         class Vector2 {
@@ -57,6 +61,9 @@
             clone() {
                 return new Vector3(this.x, this.y, this.z, this.w);
             }
+            equal(vec) {
+                return this.x == vec.x && this.y == vec.y && this.z == vec.z;
+            }
 
             static add(vec1, vec2) {
                 return new Vector3(vec1.x + vec2.x, vec1.y + vec2.y, vec1.z + vec2.z);
@@ -1028,6 +1035,36 @@
                 return allTris;
             }
 
+            setPos(pos) {
+                this.pos = pos;
+                this.quad.pos = pos;
+
+                let portalEdgeThickness = 0.2;
+                let portalBack = this.rotation.rotateVector(Vector3.back).normalize();
+                let backOffset = Vector3.mul(portalBack, portalEdgeThickness / 2);
+
+                let portalUp = this.rotation.rotateVector(Vector3.up);
+                this.topCube.pos = Vector3.add(pos, Vector3.add(Vector3.mul(portalUp, this.scale.y / 2 + portalEdgeThickness / 2), backOffset));
+                this.bottomCube.pos = Vector3.add(pos, Vector3.add(Vector3.mul(portalUp, -this.scale.y / 2 - portalEdgeThickness / 2), backOffset));
+
+                let portalRight = this.rotation.rotateVector(Vector3.right);
+                this.leftCube.pos = Vector3.add(pos, Vector3.add(Vector3.mul(portalRight, -this.scale.x / 2 - portalEdgeThickness / 2), backOffset));
+                this.rightCube.pos = Vector3.add(pos, Vector3.add(Vector3.mul(portalRight, this.scale.x / 2 + portalEdgeThickness / 2), backOffset));
+
+                this.backQuad.pos = Vector3.add(pos, Vector3.mul(portalBack, portalEdgeThickness));
+            }
+
+            rotate(eulerX, eulerY, eulerZ) {
+                let q = Quaternion.buildQuaternionEuler(new Vector3(eulerX, eulerY, eulerZ));
+                this.rotation = Quaternion.multiply(q, this.rotation);
+                this.quad.rotate(eulerX, eulerY, eulerZ);
+                this.topCube.rotate(eulerX, eulerY, eulerZ);
+                this.bottomCube.rotate(eulerX, eulerY, eulerZ);
+                this.leftCube.rotate(eulerX, eulerY, eulerZ);
+                this.rightCube.rotate(eulerX, eulerY, eulerZ);
+                this.backQuad.rotate(eulerX, eulerY, eulerZ);
+            }
+
             getPlane() {
                 let forward = this.rotation.rotateVector(Vector3.forward);
                 return new Plane(forward, Vector3.sub(this.pos, Vector3.mul(forward, 0.01)));
@@ -1146,6 +1183,8 @@
 
         const lightDirIntensitySlider = document.getElementById("dirIntensity");
         const lightDirIntensityText = document.getElementById("dirIntensityText");
+        const controllingObjectText = document.getElementById("controllingObject");
+
         let dirLightDefaultDir = new Vector3(0.2, -1, 0);
         let dirLight = new DirectionalLight(dirLightDefaultDir, parseFloat(lightDirIntensitySlider.value));
 
@@ -1156,10 +1195,15 @@
         let ground = new GameObject(new Vector3(0, 0, 0), Vector3.zero, new Vector3(30, 1, 30), new Color(0, 255, 0), cubeMesh, brickTexture);
         let playerCube = new GameObject(camera.pos, camera.rotation, new Vector3(0.5, 0.5, 0.5), new Color(255, 255, 255),
             new ObjMesh("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/models/grass_block.obj"), playerCubeTexture);
-        let portal1 = new Portal("Portal 1", new Vector3(-3, 3, -3), new Vector3(0, 180, 0), new Vector3(3, 6, 3), new Color(255, 201, 94));
-        let portal2 = new Portal("Portal 2", new Vector3(3, 3, -3), new Vector3(0, 180, 0), new Vector3(3, 6, 3), new Color(17, 155, 255));
+        let portal1Pos = new Vector3(-3, 3, -3);
+        let portal2Pos = new Vector3(3, 3, -3);
+        let portal1 = new Portal("Portal 1", portal1Pos, new Vector3(0, 180, 0), new Vector3(3, 6, 3), new Color(255, 201, 94));
+        let portal2 = new Portal("Portal 2", portal2Pos, new Vector3(0, 180, 0), new Vector3(3, 6, 3), new Color(17, 155, 255));
         portal1.setOtherPortal(portal2);
         portal2.setOtherPortal(portal1);
+        let controllingObject = 0;
+        let controllingMoveSpeed = 6;
+        let controllingRotateSpeed = 60;
 
         let lastTime = performance.now();
         let dt = 0;
@@ -1168,9 +1212,62 @@
             playerCube.pos = camera.pos;
             playerCube.rotation = camera.rotation;
 
+            let controllingPos = controllingObject == 0 ? portal1Pos : portal2Pos;
+            let posDelta = Vector3.zero;
+            if (keyStates["i"]) {
+                posDelta = Vector3.add(posDelta, Vector3.forward);
+            }
+            if (keyStates["k"]) {
+                posDelta = Vector3.add(posDelta, Vector3.back);
+            }
+            if (keyStates["j"]) {
+                posDelta = Vector3.add(posDelta, Vector3.left);
+            }
+            if (keyStates["l"]) {
+                posDelta = Vector3.add(posDelta, Vector3.right);
+            }
+            if (keyStates["u"]) {
+                posDelta = Vector3.add(posDelta, Vector3.up);
+            }
+            if (keyStates["o"]) {
+                posDelta = Vector3.add(posDelta, Vector3.down);
+            }
+            if (!posDelta.equal(Vector3.zero)) {
+                posDelta.normalize();
+                posDelta = Vector3.mul(posDelta, dt * controllingMoveSpeed);
+
+                let newPos = Vector3.add(controllingPos, posDelta);
+                controllingPos.x = newPos.x;
+                controllingPos.y = newPos.y;
+                controllingPos.z = newPos.z;
+            }
+
+            let rotateEuler = Vector3.zero.clone();
+            if (keyStates["b"]) {
+                rotateEuler.x = dt * controllingRotateSpeed;
+            }
+            if (keyStates["n"]) {
+                rotateEuler.y = dt * controllingRotateSpeed;
+            }
+            if (keyStates["m"]) {
+                rotateEuler.z = dt * controllingRotateSpeed;
+            }
+            if (!rotateEuler.equal(Vector3.zero)) {
+                switch (controllingObject) {
+                    case 0:
+                        portal1.rotate(rotateEuler.x, rotateEuler.y, rotateEuler.z);
+                        break;
+                    case 1:
+                        portal2.rotate(rotateEuler.x, rotateEuler.y, rotateEuler.z);
+                        break;
+                }
+            }
+
             camera.updateMovement(dt, keyStates);
 
+            portal1.setPos(portal1Pos);
             portal1.updatePortalCameraBaseOnPlayer(camera.pos, camera.rotation);
+            portal2.setPos(portal2Pos);
             portal2.updatePortalCameraBaseOnPlayer(camera.pos, camera.rotation);
 
             let cameraTeleportedTransform = portal1.getPlayerTeleportedTransformIfGoThroughPortal(camera.pos, camera.rotation);
@@ -1278,6 +1375,12 @@
         window.addEventListener('keyup', (e) => {
             keyStates[e.key] = false;
         });
+        window.addEventListener("keypress", (e) => {
+            if (e.key == "p") {
+                controllingObject = 1 - controllingObject;
+                controllingObjectText.innerText = controllingObject == 0 ? "Orange portal" : "Blue Portal";
+            }
+        });
     </script>
 </body>
 
