--- "a/htmlPreviews\\htmlPreview8.html"
+++ "b/htmlPreviews\\htmlPreview9.html"
@@ -974,6 +974,8 @@
                 let portalForward = this.rotation.rotateVector(Vector3.forward).normalize();
                 let portalRight = this.rotation.rotateVector(Vector3.right);
 
+                this.lastFramePlayerInFrontOfPortal = false;
+
                 this.topCube = new GameObject(
                     Vector3.add(pos, Vector3.add(Vector3.mul(portalUp, this.scale.y / 2 + portalEdgeThickness / 2), backOffset)),
                     eulerAngles,
@@ -1099,6 +1101,27 @@
                 this.camera.pos = cameraPos;
                 this.camera.rotation = cameraRotation;
             }
+
+            isPointDirectlyInFrontOrBackOfPortal(point) {
+                let worldPointRelativeToThisPortal = Vector3.sub(point, this.pos);
+                let localPointRelativeToThisPortal = this.rotation.inverseRotateVector(worldPointRelativeToThisPortal);
+
+                return Math.abs(localPointRelativeToThisPortal.x) <= this.scale.x / 2 && Math.abs(localPointRelativeToThisPortal.y) <= this.scale.y / 2;
+            }
+
+            getPlayerTeleportedTransformIfGoThroughPortal(playerCameraPos, playerCameraRotation) {
+                let plane = this.getPlane();
+                let isPlayerInFrontOfPortal = plane.isPointInFrontOfPlane(playerCameraPos);
+
+                if (this.lastFramePlayerInFrontOfPortal && !isPlayerInFrontOfPortal && this.isPointDirectlyInFrontOrBackOfPortal(playerCameraPos)) {
+                    this.lastFramePlayerInFrontOfPortal = isPlayerInFrontOfPortal;
+                    let [newPlayerPos, newPlayerRotation] = this.calculateTransformToOtherPortalTransform(playerCameraPos, playerCameraRotation);
+                    return [newPlayerPos, newPlayerRotation];
+                } else {
+                    this.lastFramePlayerInFrontOfPortal = isPlayerInFrontOfPortal;
+                    return null;
+                }
+            }
         }
 
         const canvas = document.getElementById("gameCanvas");
@@ -1109,7 +1132,7 @@
         const fov = 60;
         let znear = 0.01, zfar = 1000;
         let projectionMatrix = Mat4x4.Projection(fov, canvas.height, canvas.width, znear, zfar);
-        let camera = new Camera(new Vector3(0, 10, -10), new Vector3(-25, 0, 0), 2, 0.2);
+        let camera = new Camera(new Vector3(0, 10, -10), new Vector3(-25, 0, 0), 6, 0.2);
         let nearPlane = new Plane(Vector3.forward, new Vector3(0, 0, znear));
         let edgePlanes = [
             new Plane(Vector3.down, new Vector3(0, 1, 0)),
@@ -1150,6 +1173,18 @@
 
             portal1.updatePortalCameraBaseOnPlayer(camera.pos, camera.rotation);
             portal2.updatePortalCameraBaseOnPlayer(camera.pos, camera.rotation);
+
+            let cameraTeleportedTransform = portal1.getPlayerTeleportedTransformIfGoThroughPortal(camera.pos, camera.rotation);
+            if (cameraTeleportedTransform != null) {
+                camera.pos = cameraTeleportedTransform[0];
+                camera.rotation = cameraTeleportedTransform[1];
+            } else {
+                cameraTeleportedTransform = portal2.getPlayerTeleportedTransformIfGoThroughPortal(camera.pos, camera.rotation);
+                if (cameraTeleportedTransform != null) {
+                    camera.pos = cameraTeleportedTransform[0];
+                    camera.rotation = cameraTeleportedTransform[1];
+                }
+            }
         }
 
         function gameObjectToWorldSpaceTriangles() {
