--- "a/htmlPreviews\\htmlPreview18.html"
+++ "b/htmlPreviews\\htmlPreview19.html"
@@ -19,7 +19,6 @@
         Controlling object:
         <span id="controllingObject">Orange portal</span>
     </div>
-    <canvas id="portal1Canvas" width="400" height="300" style="border: 2px solid #aaa; background: #fff"></canvas>
     <script>
         class Vector2 {
             constructor(u = 0, v = 0, w = 1) {
@@ -679,6 +678,10 @@
                 return Mat4x4.ViewFromPosRot(this.pos, this.rotation);
             }
 
+            getForward() {
+                return this.rotation.rotateVector(Vector3.forward);
+            }
+
             updateTransformAndCorrectRoll(newPos, newRot) {
                 this.pos = newPos;
                 this.rotation = newRot;
@@ -979,6 +982,9 @@
                 this.screenBuffer = [];
                 this.depthBuffer = [];
 
+                this.pointingPos = new Vector3(0, 0, 0);
+                this.pointingPosNormal = new Vector3(0, 0, 0);
+
                 for (let i = 0; i < canvas.height; i++) {
                     let screenRow = [];
                     let depthRow = [];
@@ -1061,12 +1067,17 @@
 
                                     let pixelInvZ = tri.uv[0].w * baryCoord0 + tri.uv[1].w * baryCoord1 + tri.uv[2].w * baryCoord2;
 
-                                    let pixelWorldPos = new Vector3(
-                                        (perspectiveWorldVertices[0].x * baryCoord0 + perspectiveWorldVertices[1].x * baryCoord1 + perspectiveWorldVertices[2].x * baryCoord2),
-                                        (perspectiveWorldVertices[0].y * baryCoord0 + perspectiveWorldVertices[1].y * baryCoord1 + perspectiveWorldVertices[2].y * baryCoord2),
-                                        (perspectiveWorldVertices[0].z * baryCoord0 + perspectiveWorldVertices[1].z * baryCoord1 + perspectiveWorldVertices[2].z * baryCoord2)
-                                    );
-                                    pixelWorldPos = Vector3.div(pixelWorldPos, pixelInvZ);
+                                    if (x == screenCenter.x && y == screenCenter.y) {
+                                        let pixelWorldPos = new Vector3(
+                                            (perspectiveWorldVertices[0].x * baryCoord0 + perspectiveWorldVertices[1].x * baryCoord1 + perspectiveWorldVertices[2].x * baryCoord2),
+                                            (perspectiveWorldVertices[0].y * baryCoord0 + perspectiveWorldVertices[1].y * baryCoord1 + perspectiveWorldVertices[2].y * baryCoord2),
+                                            (perspectiveWorldVertices[0].z * baryCoord0 + perspectiveWorldVertices[1].z * baryCoord1 + perspectiveWorldVertices[2].z * baryCoord2)
+                                        );
+                                        pixelWorldPos = Vector3.div(pixelWorldPos, pixelInvZ);
+
+                                        this.pointingPos = pixelWorldPos;
+                                        this.pointingPosNormal = triWorldNormal;
+                                    }
 
                                     let lightIntensity = dirLightDiffuse;
 
@@ -1231,6 +1242,17 @@
                 this.backQuad.pos = Vector3.add(pos, Vector3.mul(portalBack, portalEdgeThickness));
             }
 
+            setRotation(newRotation) {
+                this.quad.rotation = newRotation;
+                this.topCube.rotation = newRotation;
+                this.bottomCube.rotation = newRotation;
+                this.leftCube.rotation = newRotation;
+                this.rightCube.rotation = newRotation;
+                let newUp = newRotation.rotateVector(Vector3.up);
+                this.backQuad.rotation = Quaternion.multiply(Quaternion.buildQuaternionAxisAngle(newUp, 180), newRotation);
+                this.rotation = newRotation;
+            }
+
             rotate(eulerX, eulerY, eulerZ) {
                 let q = Quaternion.buildQuaternionEuler(new Vector3(eulerX, eulerY, eulerZ));
                 this.rotation = Quaternion.multiply(q, this.rotation);
@@ -1458,9 +1480,7 @@
         }
 
         const canvas = document.getElementById("gameCanvas");
-        const portal1Canvas = document.getElementById("portal1Canvas");
         const ctx = canvas.getContext("2d");
-        const portal1Ctx = portal1Canvas.getContext("2d");
         const rasterizer = new Rasterizer();
         const fov = 60;
         let znear = 0.01, zfar = 1000;
@@ -1475,6 +1495,7 @@
             new Plane(Vector3.left, new Vector3(1, 0, 0))
         ];
         let keyStates = [];
+        let screenCenter = new Vector3(Math.floor(canvas.width / 2), Math.floor(canvas.height / 2), 0);
 
         let cameraRotLerpStart = camera.rotation;
         let cameraRotLerpEnd = camera.rotation;
@@ -1496,7 +1517,16 @@
         let brickTexture = new Texture("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/images/textures/brick.jpg");
 
         let cubeMesh = new Cube();
-        let ground = new GameObject(new Vector3(0, 0, 0), Vector3.zero, new Vector3(30, 1, 30), new Color(0, 255, 0), cubeMesh, brickTexture);
+        let sceneObjects = [
+            new GameObject(new Vector3(0, 0, 0), Vector3.zero, new Vector3(60, 1, 60), new Color(0, 255, 0), cubeMesh, brickTexture),
+            new GameObject(new Vector3(30, 10, 0), Vector3.zero, new Vector3(1, 20, 60), new Color(0, 255, 0), cubeMesh, brickTexture),
+            new GameObject(new Vector3(-30, 10, 0), Vector3.zero, new Vector3(1, 20, 60), new Color(0, 255, 0), cubeMesh, brickTexture),
+            new GameObject(new Vector3(0, 10, 30), Vector3.zero, new Vector3(60, 20, 1), new Color(0, 255, 0), cubeMesh, brickTexture),
+            new GameObject(new Vector3(0, 10, -30), Vector3.zero, new Vector3(60, 20, 1), new Color(0, 255, 0), cubeMesh, brickTexture),
+            new GameObject(new Vector3(0, 10, -3), Vector3.zero, new Vector3(8, 1, 8), new Color(0, 255, 0), cubeMesh, brickTexture),
+            new GameObject(new Vector3(-9, 5, -3), new Vector3(0, 0, -45), new Vector3(14, 1, 8), new Color(0, 255, 0), cubeMesh, brickTexture),
+            new GameObject(new Vector3(9, 5, -3), new Vector3(0, 0, 45), new Vector3(14, 1, 8), new Color(0, 255, 0), cubeMesh, brickTexture),
+        ];
         let playerCube = new GameObject(camera.pos, camera.rotation, new Vector3(0.5, 0.5, 0.5), new Color(255, 255, 255),
             new ObjMesh("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/models/grass_block.obj"), playerCubeTexture);
         let testCube = new GameObject(
@@ -1644,7 +1674,8 @@
 
         function gameObjectToWorldSpaceTriangles() {
             let objectTris = [];
-            objectTris.push(...ground.getTransformedTriangles());
+            for (let i = 0; i < sceneObjects.length; i++)
+                objectTris.push(...sceneObjects[i].getTransformedTriangles());
             objectTris.push(...portal1.getAllTransformedTriangles());
             objectTris.push(...portal2.getAllTransformedTriangles());
 
@@ -1698,11 +1729,15 @@
         }
 
         function rasterize(visibleTris) {
-            portal1.rasterizer.drawCall(portal1Ctx);
-
             rasterizer.clearScreen();
             rasterizer.rasterizeClipSpaceTriangles(visibleTris);
             rasterizer.drawCall();
+
+            let x = screenCenter.x, y = screenCenter.y;
+            ctx.fillStyle = "rgba(0, 0, 0, 0.6)";
+            ctx.fillRect(x - 2, y - 2, 5, 5);
+            ctx.fillStyle = "#ffffff";
+            ctx.fillRect(x - 1, y - 1, 3, 3);
         }
 
         function update(time = performance.now()) {
@@ -1729,6 +1764,36 @@
 
         update();
 
+        function shootPortal(portal) {
+            let cameraForward = camera.getForward();
+
+            if (portal == 0) {
+                portal1Pos = Vector3.add(rasterizer.pointingPos, Vector3.mul(rasterizer.pointingPosNormal, 0.1));
+                if (Math.abs(Vector3.dot(rasterizer.pointingPosNormal, Vector3.up)) >= 0.9) {
+                    portal1.setRotation(Quaternion.lookRotation(rasterizer.pointingPosNormal, cameraForward));
+                } else {
+                    portal1.setRotation(Quaternion.lookRotation(rasterizer.pointingPosNormal, Vector3.up));
+                }
+            } else if (portal == 1) {
+                portal2Pos = Vector3.add(rasterizer.pointingPos, Vector3.mul(rasterizer.pointingPosNormal, 0.1));
+                if (Math.abs(Vector3.dot(rasterizer.pointingPosNormal, Vector3.up)) >= 0.9) {
+                    portal2.setRotation(Quaternion.lookRotation(rasterizer.pointingPosNormal, cameraForward));
+                } else {
+                    portal2.setRotation(Quaternion.lookRotation(rasterizer.pointingPosNormal, Vector3.up));
+                }
+            }
+        }
+
+        document.addEventListener("mousedown", (e) => {
+            if (document.pointerLockElement === canvas) {
+                if (e.button == 0) {
+                    shootPortal(0);
+                } else if (e.button == 2) {
+                    shootPortal(1);
+                }
+            }
+        });
+
         lightDirIntensitySlider.addEventListener("input", () => {
             dirLight.intensity = parseFloat(lightDirIntensitySlider.value);
             lightDirIntensityText.textContent = dirLight.intensity.toFixed(2);
