--- "a/htmlPreviews\\htmlPreview2.html"
+++ "b/htmlPreviews\\htmlPreview3.html"
@@ -534,6 +534,35 @@
             }
         }
 
+        class Quad extends Mesh {
+            constructor() {
+                super(Quad.generateInitialTriangles());
+            }
+
+            static generateInitialTriangles() {
+                let tris = [];
+                let color = new Color(255, 255, 255);
+
+                tris.push(
+                    new Triangle(
+                        [new Vector3(0.5, -0.5, 0), new Vector3(0.5, 0.5, 0), new Vector3(-0.5, 0.5, 0)],
+                        [new Vector2(1, 0), new Vector2(1, 1), new Vector2(0, 1)],
+                        color,
+                    ),
+                );
+
+                tris.push(
+                    new Triangle(
+                        [new Vector3(0.5, -0.5, 0), new Vector3(-0.5, 0.5, 0), new Vector3(-0.5, -0.5, 0)],
+                        [new Vector2(1, 0), new Vector2(0, 1), new Vector2(0, 0)],
+                        color,
+                    ),
+                );
+
+                return tris;
+            }
+        }
+
         class Cube extends Mesh {
             constructor() {
                 super(Cube.generateInitialTriangles());
@@ -921,13 +950,16 @@
 
         const lightDirIntensitySlider = document.getElementById("dirIntensity");
         const lightDirIntensityText = document.getElementById("dirIntensityText");
-        let dirLightDefaultDir = new Vector3(-1, -1, -1);
+        let dirLightDefaultDir = new Vector3(0.2, -1, 0);
         let dirLight = new DirectionalLight(dirLightDefaultDir, parseFloat(lightDirIntensitySlider.value));
 
         let brickTexture = new Texture("https://raw.githubusercontent.com/hachihao792001/hachihao792001.github.io/refs/heads/main/blogs/graphics3D/images/textures/brick.jpg");
 
+        let quadMesh = new Quad();
         let cubeMesh = new Cube();
         let ground = new GameObject(new Vector3(0, 0, 0), Vector3.zero, new Vector3(30, 1, 30), new Color(0, 255, 0), cubeMesh, brickTexture);
+        let portal1 = new GameObject(new Vector3(-3, 3, 0), new Vector3(0, 180, 0), new Vector3(3, 6, 3), new Color(255, 201, 94), quadMesh);
+        let portal2 = new GameObject(new Vector3(3, 3, 0), new Vector3(0, 180, 0), new Vector3(3, 6, 3), new Color(17, 155, 255), quadMesh);
 
         let lastTime = performance.now();
         let dt = 0;
@@ -951,6 +983,8 @@
 
             let objectTris = [];
             objectTris.push(...ground.getTransformedTriangles());
+            objectTris.push(...portal1.getTransformedTriangles());
+            objectTris.push(...portal2.getTransformedTriangles());
 
             let visibleTris = [];
             for (let tri of objectTris) {
